### ISSUE TITLE
Cancel ongoing tests when new tests are requested for a PR

### ISSUE THREAD

**Issue by kpedro88** [cms-sw/cms-bot#748]
Subject line says it all...

If there is some scenario in which we want a previous round of tests to continue even though a new round has been requested, perhaps a "cancel tests" command could be added to the bot instead.

**Comment by smuzaffar** [comment]
sounds reasonable. I would say that if "please test" command is issued after a PR update then previous tests should be cancelled and restarted.

**Comment by slava77** [comment]
On 9/22/16 10:38 PM, Malik Shahzad Muzaffar wrote:
> sounds reasonable. I would say that if "please test" command is issued
> after a PR update then previous tests should be cancelled and restarted.

Maybe better not make this automated?

sometimes an update introduces a bug in a small scope
sometimes an update is almost obvious but the restarted tests fail due
to IB or other issues.
For these examples we end up without any tests or QA of a large fraction
of the PR.

So, maybe "please restart test" would be more successful?

It's much easier to ignore a set of results for an intermediate commit
than to be stuck without any results.

**Comment by smuzaffar** [comment]
Yes, killing of the previous tests is not going to be automatic on the PR update. It should only happen when a PR was updated and someone ask for the test after the update.

**Comment by slava77** [comment]
On 9/23/16 12:50 AM, Malik Shahzad Muzaffar wrote:
> Yes, killing of the previous tests is not going to be automatic on the
> PR update. It should only happen when a PR was updated and someone ask
> for the test after the update.

My suggestion was to have  different kind of ways to ask for tests
- "please test":  to keep the current behavior
- "please restart test" to restart

**Comment by smuzaffar** [comment]
current behaviour is
- "please test", starts tests and set "test running" status for PR
- on PR update, tests status is reset to "test pending". Any jenkins job keeps on running
- after the PR update, if you say "please test" then bot starts another test. So at one point there could be two jenkins jobs testing the same PR. Results of first job will be ignored as they were using a old commit hash.

So my suggestion is that why bother running two tests if the results of first one are anyway going to be ignored by bot, why not just kill previous test (if any) before starting a new one for the same PR? No need to have a different bot command for this.

**Comment by slava77** [comment]
On 9/23/16 3:29 AM, Malik Shahzad Muzaffar wrote:
> current behaviour is
> 
>   * "please test", starts tests and set "test running" status for PR
>   * on PR update, tests status is reset to "test pending". Any jenkins
>     job keeps on running
>   * after the PR update, if you say "please test" then bot starts
>     another test. So at one point there could be two jenkins jobs
>     testing the same PR. Results of first job will be ignored as they
>     were using a old commit hash.
> 
> So my suggestion is that why bother running two tests if the results of
> first one are anyway going to be ignored by bot, why not just kill
> previous test (if any) before starting a new one for the same PR? No
> need to have a different bot command for this.

I gave a couple of examples earlier in the thread, mostly driven by
usefulness of a human looking at the result rather than the bot setting
a label.
Since the success rate per PR due to infrastructure issues is below ~90%,
I consider failure rate too high and hope to keep some benefit of a
working test even if it was made before an update.
This has limited scope though, I'm not going to push for this.

I'm assuming that the results of already completed tests will not be
removed with the same logic that only the final result completion
matters for the bot label assignment.

Also, along Kevin's request,
we need syntax to restart tests (and cancel ongoing) when PR stays
unchanged but the conditions where it will be tested do (IB change,
another PR merged, something else).

**Comment by smuzaffar** [comment]
Yes, results for already finished tests will be available.

Currently when jenkins queue is big then I do kill jenkins jobs for same PR (if any) to get some resources free. I can provide "cancel test" to explicitly cancel any on going tests. But would still prefer to have only one jenkins job per PR running at a time.

**Comment by smuzaffar** [comment]
@kpedro88 , cms-bot should not allow you to abort a running/pending test. cms-bot commands are available here  https://github.com/cms-sw/cmssw/issues/13028
Developers, who can request to start a test, can now issue any of the following command to force kill an ongoing test or the test which is pending in the queue to start. 
 * abort
 * abort test
 * please abort
 * please abort test

**Comment by kpedro88** [comment]
@smuzaffar thanks! I'm sure I will have an occasion to try it out soon...

**Comment by kpedro88** [comment]
@smuzaffar I tried this on https://github.com/cms-sw/cmssw/pull/16670, which had tests already finished, but the comparison job was still queued (hadn't run yet). Does the abort functionality also kill the comparison job?

**Comment by smuzaffar** [comment]
I think not. It only recognize the abort command if tests are still running (means bot has not yet added a +1 or -1). I guess I should change it to also recognize it if comparison is also on-groing

**Comment by smuzaffar** [comment]
@kpedro88 , by the way if a PR is updated and someone asks for "please test" then the pr testing job will actually kill any running tests for this PR (including comparison tests). So in this case, no need to explicitly ask for abort.

**Comment by kpedro88** [comment]
@smuzaffar in https://github.com/cms-sw/cmssw/pull/16670, I asked "please test" while a comparison job had been queued from the previous test. 15 minutes later, that comparison result (from a previous commit) was posted... so I don't think this automatic cancellation works yet, unless I misunderstand what it's supposed to do.

**Comment by smuzaffar** [comment]
there was typo in the jenkins pr test job name that is why it did not started the stop-ib-any-integration job. This should be fixed not (I hope, jenkins is very sensitive with space characters :-) )

**Comment by smuzaffar** [comment]
looks like test abort command is working, so closing this issue. Please re-open or create a new issue if "test abort" or "please test" is not killing the previously running tests.
